File: solutions/haskell/state-of-tic-tac-toe/1/src/StateOfTicTacToe.hs
module StateOfTicTacToe (gameState, GameState (..)) where

data GameState = WinX | WinO | Draw | Ongoing | Impossible deriving (Eq, Show)

gameState :: [String] -> GameState
gameState board
  | countX > 1 && countO == 0
      || countO > countX
      || win 'X' && (countX <= countO)
      || win 'O' && (countO < countX) =
      Impossible
  | win 'X' = WinX
  | win 'O' = WinO
  | ' ' `notElem` flat = Draw
  | otherwise = Ongoing
 where
  flat = concat board
  countX = length [x | x <- flat, x == 'X']
  countO = length [x | x <- flat, x == 'O']
  win p =
    True
      `elem` [ (p, p, p) == (flat !! x, flat !! y, flat !! z)
             | (x, y, z) <-
                [ (0, 1, 2) -- lines
                , (3, 4, 5)
                , (6, 7, 8)
                , (0, 3, 6) -- cols
                , (1, 4, 7)
                , (2, 5, 8)
                , (0, 4, 8) -- diags
                , (2, 4, 6)
                ]
             ]

File: solutions/lua/state-of-tic-tac-toe/1/state-of-tic-tac-toe.lua
local function win(s, v)
	local positions = {
		("^%s%s%s......"):format(v, v, v), -- lines (1,2,3)
		("^...%s%s%s..."):format(v, v, v), --    (4,5,6)
		("^......%s%s%s"):format(v, v, v), -- (7,8,9)
		("^%s..%s..%s.."):format(v, v, v), -- cols (1,4,7)
		("^.%s..%s..%s."):format(v, v, v), --      (2,5,8)
		("^..%s..%s..%s"):format(v, v, v), --      (3,6,9)
		("^%s...%s...%s"):format(v, v, v), -- diags (1,5,9)
		("^..%s.%s.%s.."):format(v, v, v), --       (3,5,7)
	}
	for _, p in ipairs(positions) do -- any match
		if s:match(p) then return true end
	end
	return false
end
return {
	gamestate = function(board)
		local flat = table.concat(board)
		local _, countX = flat:gsub("X", "")
		local _, countO = flat:gsub("O", "")
		local winX = win(flat, "X")
		local winO = win(flat, "O")

		assert(countX >= countO)
		if countX > 1 then assert(countO > 0) end
		if winX then assert(countX > countO) end
		if winO then assert(countO >= countX) end

		if winX or winO then return "win" end
		if not flat:match(" ") then return "draw" end
		return "ongoing"
	end,
}
